File: Antlr/src/main/scala/io/github/chaosunity/casc/util/ReflectionMapper.scala
package io.github.chaosunity.casc.util

import io.github.chaosunity.casc.parsing.`type`.ClassType
import io.github.chaosunity.casc.parsing.expression.FunctionParameter
import io.github.chaosunity.casc.parsing.scope.FunctionSignature

import java.lang.reflect.{Constructor, Executable, Method, Parameter}
import java.util
import java.util.Arrays.asList

object ReflectionMapper {
    def fromMethod(method: Method): FunctionSignature = {
        val (name, parameters) = fromExecutable(method)
        val returnType = method.getReturnType

        new FunctionSignature(name, parameters, new ClassType(returnType.getCanonicalName))
    }

    def fromConstructor(constructor: Constructor[_]): FunctionSignature = {
        val (name, parameters) = fromExecutable(constructor)
        val returnType = constructor.getDeclaringClass

        new FunctionSignature(name, parameters, new ClassType(returnType.getCanonicalName))
    }

    private def fromExecutable(executable: Executable): (String, util.List[FunctionParameter]) =
        (executable.getName, asList(executable.getParameters.map((param: Parameter) =>
            new FunctionParameter(new ClassType(param.getType.getCanonicalName), param.getName, Option.empty, false)
        )))
}

File: src/main/kotlin/io/github/chaosunity/casc/util/ReflectionMapper.kt
package io.github.chaosunity.casc.util

import io.github.chaosunity.casc.parsing.node.expression.Parameter
import io.github.chaosunity.casc.parsing.scope.FunctionSignature
import io.github.chaosunity.casc.parsing.type.BuiltInType
import java.lang.reflect.Constructor
import java.lang.reflect.Executable
import java.lang.reflect.Method
import java.lang.reflect.Modifier

object ReflectionMapper {
    fun fromMethod(method: Method): FunctionSignature {
        val (name, parameters) = fromExecutable(method)
        val returnTypeClass = method.returnType

        return FunctionSignature(
            name,
            parameters,
            TypeResolver.getTypeByName(returnTypeClass.typeName),
            Modifier.isStatic(method.modifiers)
        )
    }

    fun fromConstructor(constructor: Constructor<*>): FunctionSignature {
        val (name, parameters) = fromExecutable(constructor)

        return FunctionSignature(name, parameters, BuiltInType.VOID, false)
    }

    private fun fromExecutable(executable: Executable): Pair<String, List<Parameter>> =
        executable.name to executable.parameters.map {
            Parameter(
                TypeResolver.getTypeByName(it.type.canonicalName),
                it.name,
                null
            )
        }
}